libass/ass_utils.h

```diff
@@ -182,19 +183,37 @@ static inline int double_to_d22(double x)
     return lrint(x * 0x400000);
 }
 
+static inline int double_to_int(double d)
+{
+    // For the usual 64bit double, 32bit int this
+    // differs for exactly INT_MAX.
+    // For other sizes (int) ((double) INT_MAX) may be UB
+    // if INT_MAX is not accurately representable as a double
```

Comment:
This comment is inaccurate: if `((double)INT_MAX)` is not accurately representable, it's rounded in an implementation-defined manner. It's only UB if `INT_MAX` is out of range for `double` _and_ IEEE arithmetic isn't supported. We only really support 64-bit-double anyway, though, so it's all kinda moot.

Reply:
Also, couldn't we use `[l]lrint` or `[l]lround` here, and then return `INT_MIN`/`LLONG_MIN` on `fetestexcept(FE_INVALID)`? This would only work on IEEE platforms, but… do we care about non-IEEE platforms?

I looked through every platform on godbolt and the only ones without `math_errhandling & MATH_ERREXCEPT` I found were: 
- AVR
- KVX
- MRISC32
- MSP430
- RISC-V rv32gc
- ESP32 (inc. S2/S3)

Out of these platforms, only rv32gc seems to be a plausible target for libass, and as far as I can tell, it's only missing this because of lack of existing implementation support (the hardware supports it).

We could also use these checks and return `INT_MIN` in cases where we're sure the result will be incorrect, and otherwise use `lrint`/`lround` to handle the straggler cases in a manner that's portable for conformant IEEE platforms, and otherwise may return some edge-case value (but is never UB).

Reply:
The `(int)` is not meant to be a comment about what type can have different sizes, but part of the expression, so the full expression is `(int) ((double) INT_MAX)` with two casts.

The first one from an integer to a floating-point type is to my understanding¹ indeed not UB as you pointed out if absolutely larger regular numbers or ±Infinity exists *(e.g. IEEE)*.
The comment was about the second cast, from a floating-point to integer type, though which will be UB if `*_MAX` is not accurately representable and the implementation-defined rounding to the available precision adjusts the value away from zero. Then the value of `(double) *_MAX` is larger than the original `INT_MAX` value and ISO C99, section 6.1.3.4 makes it very clear that casting a float type whose integral part cannot be represented in the targeted integer representation is undefined behaviour *(yes, even `(unsigned) -2.4`)*. That's why the check has to be `>=` and not `>` and in fact this does happen on my machine for the `long long` version of this function.

----

¹: The exact wording in section 6.1.3.4 is: 
> “If the value being converted is in the range of values that can be represented but cannot be represented exactly, the result is either the nearest higher or nearest lower representable value, chosen in an implementation-defined manner. If the value being converted is outside the range of values that can be represented, the behavior is undefined.”

I interpret this to mean that if an integer value is larger than the largest regular number the floating type can represent it can still be "represented" as `INFINITY` if it exists, though I haven't checked the rest of the spec to closely about possible restrictions/exceptions about `±INFINITY`. If that’s correct, then only something like ARM's alternative 16bit float mode could have UB in the first cast; assuming IEEE float seems reasonable to me though.

Reply:
For `int32_t`, this already exists in `ass_parse.c` as `dtoi32` since #261. And it handles `INT(32)_MAX` _without_ deviating from x86 by using `INT32_MAX + 1LL`. Other than that, it’s identical to your code (notably with equally-strict comparison signs), which is reassuring.

Reply:
Switched all `argtoi` usage to argtoi32; I'm assuming this should also improve VSFilter-compatibility in some edge cases unless VSF sometimes uses a function which first parses an int as a double like `argtoi` did.

Reply:
> unless VSF sometimes uses a function which first parses an int as a double like `argtoi` did.

No, FWIW I don’t _think_ it ever does this, and I distinctly remember this being one of our longstanding quirks. But I’m not quite sure why I didn’t replace those calls immediately when I introduced `argtoi32`. Maybe it’s worth giving these instances another quick cross-check with VSFilter source code just in case.

Reply:
From quick glance a guliverkli2, all paramters replaced here and the `fade` paramters already replaced in #261 are using `wcstol` and afaik in all current Microsoft platforms `long` is 32bit.